Added video preview.

dsjen opened the pull request:
I'm still waiting on product feedback, in particular the "View Live" button, which doesn't look good in its current placement.  Please feel free to review with that in mind.

@dan-f

dsjen commented on analytics_dashboard/courses/templates/courses/_video_module_controls.html:
```diff
@@ -1,32 +1,57 @@
+    <div class="col-xs-12 collapsible-content">
+
+        {% if show_video_preview and view_live_url%}
+            {# TODO: refactor this with the previous? -- needs Product input on look of "button" #}
```
I'm waiting on product feedback on the location of the button and will refactor this depending on the decision.

dan-f commented on acceptance_tests/test_course_engagement.py:
```diff
@@ -223,6 +223,22 @@ def setUp(self):
+    def _test_video_controls(self):
+        preview_selector = '#module-preview'
+        element = self.page.q(css=preview_selector)
+        self.assertFalse(element.visible)
```
How about inlining the `self.page.q(css=preview_selector)` call here? That way `element` doesn't have to represent two separate elements.

dsjen replied:
Thanks, good idea!

dsjen commented:
@dan-f I will need to update the templates in order to meet changes product changes.  I'll let you know when that's complete.

dsjen commented:
@dan-f I've changed the layout a bit to meet updated product requirements.  Please review.

dan-f commented on acceptance_tests/test_course_engagement.py:
```diff
@@ -223,6 +223,21 @@ def setUp(self):
+        self.assertTrue(toggle_element.present)
+        self.assertEqual(toggle_element.text[0], 'EXPAND PREVIEW')
+        toggle_element.click()
+        self.assertEqual(toggle_element.text[0], 'COLLAPSE PREVIEW')
```
What happens in the bok choy test when the expand preview is clicked? Is there an LMS against which we're integrating, or is it a mock service?

dsjen replied:
We're not checking that anything is actually displayed (via LMS), but that the toggle text changes and the section is visible.

dan-f replied:
And what actually does happen? We should make sure that the behavior is deterministic so that we don't see flaky tests down the line.

dan-f commented on analytics_dashboard/courses/presenters/__init__.py:
```diff
@@ -368,3 +368,9 @@ def build_view_live_url(self, base_url, module_id):
             view_live_url = u'{0}/{1}/jump_to/{2}'.format(base_url, self.course_id, module_id)
         return view_live_url
+
+    def build_render_xblock_url(self, base_url, module_id):
```
I recommend using [`urllib.urljoin`](https://docs.python.org/2/library/urlparse.html#urlparse.urljoin) here, as it will handle leading/trailing slash edge cases.

dsjen replied:
I just tried using `urljoin`, but it didn't work because the module_id can be in a form that looks like a url (e.g. i4x://edx/demo/video/79a975af9e1f4a4a99b21c49aa3908c4) and `urljoin` ends up resolving the url to `i4x://...` in that case.

I like the idea of using `urljoin`.  Is there something I'm missing?

dsjen replied:
I added a helper method `_build_url` to take care of the slashes.

dan-f replied:
Nice, looks great!

dan-f commented on analytics_dashboard/courses/tests/test_presenters.py:
```diff
@@ -546,6 +546,13 @@ def build_module_url_func(parent, video):
+    def test_build_render_xblock_url(self):
+        self.assertIsNone(self.presenter.build_render_xblock_url(None, None))
+
+        xblock_render_base = 'http://example.com'
```
Test when base URL has a trailing slash and no trailing slash :)

dsjen replied:
Added ddt for trailing slash.

dan-f replied:
:heart: ddt

dan-f commented on analytics_dashboard/settings/base.py:
```diff
@@ -420,3 +420,7 @@
 ########## END COURSE API
+
+########## MODULE_PREVIEW
+MODULE_PREVIEW_URL = 'http://127.0.0.1:8000/xblock'
```
I'm not sure that this `127.0.0.1` value should go in `base.py`, since `production.py` inherits from it. How about adding it to `dev.py`? This seems like a setting that should be configurable, so I'd recommend checking with Ops to see if we should add this to the configuration project.

dsjen replied:
Good call!  I'll set this to `None` in `base.py` and then to `127.0.0.1` in `dev.py`.  I'll let ops know that this should be set for stage, edge, and prod.

dan-f replied:
Don't we need to explicitly scrape this setting from the yaml config, then?

dan-f commented on analytics_dashboard/static/js/config.js:
```diff
@@ -29,7 +29,8 @@ require.config({
         globalize: 'bower_components/globalize/dist/globalize',
-        globalization: 'js/utils/globalization'
+        globalization: 'js/utils/globalization',
+        collapsible: 'bower_components/edx-ui-toolkit/components/views/collapsible-view'
```
Nice!

dan-f commented on analytics_dashboard/courses/templates/courses/_video_module_controls.html:
```diff
@@ -1,32 +1,19 @@
-            </div>
-        </div>
+<div class="container-fluid">
+    <div class="col-xs-6">
```
From the [bootstrap css page](http://getbootstrap.com/css/#grid-intro) I got the sense that columns should be parented by rows (second two bullets on that list), though it's sort of vague. Does this change styling at all?

dsjen replied:
I went ahead and wrapped this in a row--didn't change anything, but seemed good to do when nesting in a container.

dan-f commented on analytics_dashboard/courses/templates/courses/_video_preview.html:
```diff
@@ -0,0 +1,28 @@
+
+<div class="module-preview-collapsible container-fluid" data-collapse-text="{{ collapse_text }}" data-expand-text="{{ expand_text }}">
+
+    <div class="col-xs-12">
```
Wondering (as above) if this needs to be wrapped in a row div.

dsjen replied:
Add row here.

dan-f commented on analytics_dashboard/courses/templates/courses/_video_preview.html:
```diff
@@ -0,0 +1,28 @@
+    <div class="col-xs-12">
+        <span class="pull-right">
+            <a href="{{ view_live_url }}" class="btn btn-default" role="button" target="_blank">
+                {# Translators: This text will be a direct link to the a specific module/problem.#}
```
Obsessive nit: add a space between end of comment and the `#`

dsjen replied:
Done. :)

dan-f commented on analytics_dashboard/courses/templates/courses/_video_preview.html:
```diff
@@ -0,0 +1,28 @@
+        {% endif %}
+    </div>
+
+    <div class="col-xs-12 collapsible-content">
```
Still confused about rows. Since this content appears beneath the previous full-width column, shouldn't this be a row? Would there ever be a scenario in which these columns accidentally ended up inline?

dsjen replied:
I'm going to leave this one w/o a row because it's already wrapped in a row above now.  They'll be on separate lines at all times because nothing is smaller than an `xs` and both are taking up full-width as you noted.

dan-f commented on analytics_dashboard/courses/templates/courses/_video_preview.html:
```diff
@@ -0,0 +1,28 @@
+{% load i18n %}
+
+<div class="module-preview-collapsible container-fluid" data-collapse-text="{{ collapse_text }}" data-expand-text="{{ expand_text }}">
```
I'm not crazy about how this div has three attributes related to collapse/expand regardless of whether the feature is enabled. It also is slightly confusing that the view live link is wrapped up in the collapse/expand component. Is there a markup structure that could more clearly separate the concerns of these components?

dsjen replied:
This is a good point.  I can add a check in the parent page to route to either the original `_module_description.html` or `_video_preview.html` and remove some of the if statements in this template.

cptvitamin replied:
don't forget ```aria-expanded="[true|false]"``` That is the one you should be using ;)  http://www.w3.org/TR/wai-aria-1.1/#aria-expanded

dsjen replied:
Thanks!  I've added in `aria-expanded="false" aria-controls="collapsible-preview"` attributes to the button, but will need to make a PR against the edx-ui-toolkit in order to toggle `aria-expanded` appropriately.

dan-f commented on analytics_dashboard/courses/templates/courses/_video_preview.html:
```diff
@@ -0,0 +1,28 @@
+            </a>
+        </span>
+        {% if show_video_preview %}
+            <div class="collapsible-toggle-text">
```
Since this is a clickable action, it should probably be a `button` for accessibility reasons, right? cc @cptvitamin

dsjen replied:
Good point!  I've updated to a button and confirmed the change w/ product.

dan-f commented on analytics_dashboard/static/js/views/iframe-view.js:
```diff
@@ -0,0 +1,29 @@
+                self.$el.on('load', function () {
+                    $(self.options.loadingSelector).remove();
+                });
+            }
```
Add `return this;`

dsjen replied:
Updated.

dan-f commented on analytics_dashboard/static/js/views/iframe-view.js:
```diff
@@ -0,0 +1,29 @@
+
+        render: function () {
+            var self = this;
+            self.$el.attr('src', self.$el.data('src'));
```
Why not do this on the server-side template? Is this what initiates the loading of the iframe's content?

dsjen replied:
We could set the `src` in the django template, but then the iframe loading becomes part of the critical path, holding up JS from kicking off.  This line is basically the key feature of this module--to load the iframe source afterwards.

dan-f replied:
That's what I was hoping. Nifty module!

dsjen replied:
Thanks! :dancers:

dan-f commented on analytics_dashboard/static/js/views/iframe-view.js:
```diff
@@ -0,0 +1,29 @@
+
+        initialize: function(options) {
+            this.options = options;
+            this.render();
```
The TNL team tended to not call `render` from `initialize` as a general practice. I don't see anything wrong with it here, but it can be useful in applications that display multiple views and don't want to necessarily render all of them.

dan-f commented on analytics_dashboard/static/js/views/iframe-view.js:
```diff
@@ -0,0 +1,29 @@
+        render: function () {
+            var self = this;
+            self.$el.attr('src', self.$el.data('src'));
+            if (self.options.loadingSelector) {
```
Could this view own its own loading indicator? I can understand wanting the loading indicator to be rendered immediately. Is this script optimized in a bundle that's served on the first page load, or does it come in a subsequent request?

dsjen replied:
Yes, the scripts are optimized and bundled for speedy delivery.  I guess the view could own the loading entirely, but it seems a bit more flexible if the spinner was provided as an option.

dan-f commented:
@dsjen I finished my review. A few comments and many questions. Things look really good!

dsjen commented:
@dan-f I've pushed a new version and addressed your comments.  Please let me know if there's anything else.

A follow up PR against the edx-ui-toolkit is needed for updating the aria-expanded attribute.

dan-f commented on acceptance_tests/test_course_engagement.py:
```diff
@@ -223,6 +223,21 @@ def setUp(self):
+        toggle_element = self.page.q(css='.collapsible-toggle-text')
+        self.assertTrue(toggle_element.present)
+        self.assertEqual(toggle_element.text[0], 'Expand Preview')
+        toggle_element.click()
```
@dsjen I made a comment on a thread but it was folded by the diff, so I'm not sure if you got the notification - I'll post it below:

(with regards to what happens to the iframe) "...what actually does happen? We should make sure that the behavior is deterministic so that we don't see flaky tests down the line."

dsjen replied:
Thanks for checking!  The click doesn't load the frame--just expands the preview section and then the subsequent lines will check to see if it's expanded (not loaded necessarily).

dan-f commented:
:+1: other than my remaining question about the bok choy test.

dan-f commented:
:+1: :shipit:
